chapters/programming_opencl_kernels.md

```diff
@@ -16,16 +16,25 @@ The OpenCL specification also enables optional *offline* compilation where the k
 
 To offset this portability problem, and to enable a richer language and compiler ecosystem, Khronos has defined a cross-vendor, portable intermediate program representation called [SPIR-V](https://www.khronos.org/spir/). An increasing number of OpenCL implementations are supporting ingestion of offline-compiled kernel programs in the SPIR-V format.
 
+SPIR-V enables independent innovation by the compiler and silicon communities. Compiler front ends that generate SPIR-V kernels that can be ingested and executed by any OpenCL driver that understands the SPIR-V format. For example the  [C++ for OpenCL](cpp_for_opencl.md) open source front-end and compilers for SYCL can generate SPIR-V code.
+
+
 <p align="center">
 <br>
-<img src="../images/opencl_and_spirv.jpg" width=800 >
+<img src="../images/online_vs_offline_comp.jpg">
 <br> <br>
-  <b>OpenCL Language Ecosystem Enabled With SPIR-V</b>
+  <b>Online vs Offline Compilation of Kernels</b>
 <br> <br>
 </p>
 
-SPIR-V enables independent innovation by the compiler and silicon communities. Compiler front ends that generate SPIR-V kernels that can be ingested and executed by any OpenCL driver that understands the SPIR-V format. For example the  [C++ for OpenCL](https://github.com/KhronosGroup/Khronosdotorg/blob/master/api/opencl/assets/CXX_for_OpenCL.pdf) open source front-end and compilers for SYCL can generate SPIR-V code.
-
 SPIR-V also enables OpenCL kernels written in OpenCL C and C++ for OpenCL to be executed by runtimes other than OpenCL, providing more deployment flexibility for developers that have invested in OpenCL kernel programming. For example, the Google [clspv](https://github.com/google/clspv) open source compiler can generate Vulkan SPIR-V shaders from OpenCL C kernel source code, Microsoft is working on a compiler chain that can ingest [OpenCL SPIR_V kernels into DX12](https://github.com/microsoft/OpenCLOn12), and there is early work on using SPIR-V tools such as SPIRV-Cross to bring OpenCL SPIR-V kernels into Metal for execution on Apple platforms.
```

Comment:
While the change in isolation makes sense, (not having read the document previously helps in this regard) there are two things (IMHO) that could be improved, one outside the actual change set, but as a consequence of the new content:
- Microsoft's toolchain does indeed use SPIR-V compute as an intermediary step which makes it "trivial" for them to ingest it, skipping previous compilation steps, however they have not enabled it. I too have requested this addition, as it would be a vendor-agnostic way to consume SPIR-V on Windows platforms, something you can't pull off with either NV or AMD HW today. (It would also vendor agnostically enable SYCL on Windows using ComputeCpp, so the value is great for little work.) It's on their backlog, but it may never materialize. I'd chose my words more carefull than "working on a compiler chain that _can_ ingest..."
- The comment/change is about the image and a previous paragraph: "The OpenCL specification also enables optional _offline_..."
  - The diagram depicts this mode via `clGetProramInfo` which may be pointed out in text.
  - While it isn't portable, it wasn't meant to be. It's primary purpose is to let applications cache their kernel binaries, something which is automatic in the graphics world and is done by drivers. OpenCL gives this ability into the programmers hands. (Libraries like clFFT automatically take care of this and dump the binary for each device into the current working directory.) Even the spec says this:
> OpenCL allows applications to create a program object using the program source or binary and build appropriate program executables. This can be very useful as it allows applications to load program source and then compile and link to generate a program executable online on its first instance for appropriate OpenCL devices in the system. These executables can now be queried and cached by the application. Future instances of the application launching will no longer need to compile and link the program executables. The cached executables can be read and loaded by 
the application, which can help significantly reduce the application initialization time.
  - Given how it's purpose was to allow apps to cache their binaries (and not for authoring portable compilation steps) I'd change the previous paragraph to something like:

The OpenCL specification also enables (primarily for caching purposes) extracting the result of online compilation, the machine binary format that a particular driver can ingest. Because OpenCL implementations may use proprietary (potentially driver version dependent) binary formats, this mechanism cannot double as a portable way of compiling OpenCL C offline.

Reply:
> I'd chose my words more carefull than "working on a compiler chain that can ingest..."

Ok, this is not introduced in this change but if you have any good suggestions how we should change this I am happy to reword it.

>The OpenCL specification also enables (primarily for caching purposes) extracting the result of online compilation, the machine binary format that a particular driver can ingest. Because OpenCL implementations may use proprietary (potentially driver version dependent) binary formats, this mechanism cannot double as a portable way of compiling OpenCL C offline.

Ok, I guess you are referring to the following paragraph.

>The OpenCL specification also enables optional *offline* compilation where the kernel program is compiled into a machine binary format that a particular driver can ingest. OpenCL implementations may use proprietary binary formats - and so offline compilation can be less portable than using online compilation of OpenCL C.
?

Ok, I am happy to change it too. I think we should aim for something that explains that there are 2 different offline compilation flavors:
1. Uses implementation-specific non-portable binaries primarily for caching purposes that are mainly obtained during online compilation using _clGetProgramInfo_ but can also be compiled offline.
2. Uses portable binary formats primarily to mitigate compilation speed bottleneck or facilitate the use of a richer variety of languages/tools available in the ecosystem i.e. C++ for OpenCL, SYCL, etc.

Does it make sense?

Reply:
FYI I have to say that caching is also done for compilation speed reduction I guess perhaps 1 and 2 are not that different after all aside from the use of other languages in the ecosystem and portability among versions of driver or even vendirs...

Reply:
I have updated the description with an explanation about binary caching. Let me know if you would like me to add or amend anything. :)

Reply:
I will resolve this for now as any amendment can be done in the follow up PRs.